### PR TITLE
Implemented PVC stats aggregation. Exposed PVC stats through Heapster API.

### PR DESCRIPTION
**Pull request by verult**
Introducing the PVC resource, along with node, namespace, pod, and container.

Added new API endpoints `/api/v1/model/namespaces/{namespace-name}/pvcs/...`

This allows users to obtain PVC usage stats independent of pods.

### CODE REVIEW

**Comment by verult** [comment]
/test continuous-integration/travis-ci/pr

Looks like the event manager test is flaky. It happened in this PR as well: https://github.com/kubernetes/heapster/pull/1820

**Comment by verult** [comment]
Is it necessary to keep PVC stats as LabeledMetrics? It's convenient, since I can reuse `decodeFsStats`, but AFAIU the labels are used to distinguish between volumes, and that's not necessary here.

File: deploy/kube-config/influxdb/heapster.yaml [hunk 1]
```diff
@@ -24,7 +24,7 @@ spec:
         imagePullPolicy: IfNotPresent
         command:
         - /heapster
-        - --source=kubernetes:https://kubernetes.default
+        - --source=kubernetes.summary_api:''
```
Comment (DirectXMan12) [inline deploy/kube-config/influxdb/heapster.yaml]: please don't do unrelated changes in the same commit.

File: metrics/api/v1/model_handlers.go [hunk 2]
```diff
@@ -60,6 +60,10 @@ type clusterMetricsFetcher interface {
 	availablePodMetrics(request *restful.Request, response *restful.Response)
 	podMetrics(request *restful.Request, response *restful.Response)
 
+	namespacePVCList(request *restful.Request, response *restful.Response)
```
Comment (DirectXMan12) [inline metrics/api/v1/model_handlers.go]: I don't think this is necessary.  The model API is more-or-less deprecated (see #1898), and we don't really want to encourage its use by adding new stuff to it.

File: metrics/sources/summary/summary.go [hunk 3]
```diff
@@ -233,6 +240,64 @@ func (this *summaryMetricsSource) decodeContainerStats(podLabels map[string]stri
...
+				CollectionStartTime: pod.StartTime.Time,
+				ScrapeTime:          vol.Time.Time,
+			}
+
+			this.decodeFsStats(pvcMetrics, PVCResourcePrefix+pvc.Name, &vol.FsStats)
+			metrics[pvcMetricName] = pvcMetrics
+
+		} else {
+			labels := metrics[pvcMetricName].Labels
+			labels[LabelPodRefs.Key] = appendPodRef(labels[LabelPodRefs.Key], pod.PodRef.Name)
```
Comment (DirectXMan12) [inline metrics/sources/summary/summary.go]: nope, don't do this.  Leave it broken up per-pod.  People can aggregate in their sinks if they want it aggregated, but it's difficult-to-impossible to "unaggregate"

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
Nix the model change -- it's not necessary,
and you shouldn't be aggregating what amounts to a per-pod metric all together.  People can aggregate in their sinks if they want the sum for all pods for a single PVC, but you can't unaggregate.

File: deploy/kube-config/influxdb/heapster.yaml (hunk 1 above)
Comment (verult) [reply deploy/kube-config/influxdb/heapster.yaml]: Is the kubelet source still used? If not, I'll add this in another PR.

File: metrics/sources/summary/summary.go [hunk 4]
```diff
@@ -234,6 +241,64 @@ func (this *summaryMetricsSource) decodeContainerStats(podLabels map[string]stri
 	return containerMetrics
 }
 
+// decodePVCStats computes PVC metrics using volume stats of each pod.
+// It selects the volume stats of the pod with the latest scrape time.
+// Metric collection time is the earliest pod start time among all pods.
+// PVC stats are only available when the PVC is in use by at least one pod.
```
Comment (kawych) [inline metrics/sources/summary/summary.go]: Did you investigate a possibility of retrieving PVCs not used by any pod? Otherwise, what value do these metrics add? Heapster already collects (and exports) volume metrics on pod level.

Comment (verult) [reply metrics/sources/summary/summary.go]: I think it'd be a lot more difficult to obtain filesystem stats of volumes not mounted on any pod. The main value is to allow users to query PVC stats by PVC name, without having to scan through pods themselves. But maybe adding a `pvc_label` to existing volume metrics is sufficient for this purpose?

Comment (kawych) [reply metrics/sources/summary/summary.go]: I think Heapster can retrieve PVCs from api server, using core v1 lister: https://github.com/kubernetes/client-go/blob/master/listers/core/v1/persistentvolumeclaim.go. Similar examples are present in processors, for example pod_based_enricher lists all pods and retrieves some information for them: https://github.com/kubernetes/heapster/blob/master/metrics/processors/pod_based_enricher.go

I prefer this over adding pvc label to existing volume metrics, because this label would be applicable only to particular type of the volume. This would leave us without the association between PVCs and pods, but do we need it?

Comment (kawych) [reply metrics/sources/summary/summary.go]: Actually, I missed a part of it (api server don't have most data about the volumes that we need). So this approach LGTM, this comment is resolved.

**Comment by verult** [comment]
@DirectXMan12 one benefit of pre-aggregating PVC stats and treating it as a separate resource is we can now include PVC-specific information, such as storage classes and volume mode (filesystem or block). These shouldn't be included as pod-level volume metric labels because not all volumes are PVCs.

**Comment by DirectXMan12** [comment]
Heapster is now deprecated, so no new features or sinks will be added.  Please see https://github.com/kubernetes/heapster/blob/master/docs/deprecation.md#initial-deprecation-kubernetes-111 for more information.
